### PR TITLE
Snyk scanning & reporting automated workflow

### PR DESCRIPTION
**Pull request by samzsam**
### Two workflow Jobs

1. Runs during PR creation on main and outputs the findings in GitHub Workflow console
2. Runs on Cron and creates JIRA ticket based on findings

### Here is a sample JIRA output 

https://qmacbis.atlassian.net/browse/MDCT-2579

### JIRA ticket behavior 

1. Exisitng ticket check is set to default 60days, this can be adjusted as needed.
2. The same Vulnerability can re-exist after the resoultion when a the same package is added for this reason there is acheck for Closed or Cancelled ticket and based on that a ticket get created.

### SNYK Ignore 

There is a .snyk file in the root directory that can hold a list of snyk ignore ids, a vulnerabiity can be manually added there following the same synaxt in the file, so when Snyk scan runs all the vulnerabilities in .snyk file will be ignored.

### Related ticket(s)

---
### Important updates

---
### Author checklist

- [ ] I have performed a self-review of my code
- [ ] I have added [thorough](https://bit.ly/3zPrxuZ) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

**Review by oweingrod-coforma — CHANGES_REQUESTED** [review]
Thanks for this @samzsam! Code looks good, but is it possible to tighten up the output more? The linked issue is very hard to read and the formatting doesn't look like it was copied correctly. Aside from that, this looks good to me!

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:133]: Blocking: The `request-promise` library was [deprecated three years ago](https://www.npmjs.com/package/request-promise). If we can use the jira API directly, then we won't need `rp` at all. Otherwise, we should use an alternative library. `axios` and `node-fetch` seem promising. In fact it looks like you already tried the latter?

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:165]: Request: This logic is redundant. If we've already filtered the vulnerabilities to be unique by title, then there is no need to keep track of the `seenTitles`. Let's do one or the other.

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:108]: Request: This try-catch is redundant, as is the try-catch around the attachment request below. This function is already being called from within a try-catch, so let's just let the exceptions bubble up to be logged in the caller.

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:5]: Nitpick: Both `fetch` and `base64` are currently unused. Let's delete them.

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:61]: Idea: JQL supports relative date math; instead of doing the calculation and date formatting ourselves we can include the search clause `AND created >= startOfDay("-60d")`

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:71]: Idea: JQL supports status conditions. Instead of checking issue status here we could include the search clause `AND status NOT IN ("Closed", "Cancelled")`.

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:72]: Comment: JQL's string matching functionality is pretty imprecise. We should probably use a stricter check here, on the entire issue summary, to see if it exactly matches the summary of the issue we're about to create.

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:65]: Comment: We should probably escape any double quotes that may appear in the vulnerability title. We could do this with a `.replaceAll("\"", "\\\"")`

File: parse_snyk_output.cjs
Comment (benmartin-coforma) [inline parse_snyk_output.cjs:1]: Question: Would it be prudent to put this file in a subdirectory? Perhaps within .github/workflows, since it is only likely to be called from there?

**Review by benmartin-coforma — COMMENTED** [review]
I've got a lot of comments here, and I believe at least some of them should be addressed before merging. But I'm about to be out of the office for a week, so I won't block the PR myself; the team can call the shots.

File: parse_snyk_output.cjs
Comment (oweingrod-coforma) [reply parse_snyk_output.cjs:1]: This is also a good idea—can we move this to the `.github` folder?

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs]: In  JQL, there is no out-of-the-box way to escape special characters in the text-search clause. I tried the above it throws error.

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs]: Removed!

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs]: Addressed!

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs]: Addressed!

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs:1]: Done!

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs]: Done!

File: parse_snyk_output.cjs
Comment (samzsam) [reply parse_snyk_output.cjs]: Done!

File: .github/workflows/parse_snyk_output.cjs
Comment (oweingrod-coforma) [inline .github/workflows/parse_snyk_output.cjs:58]: Can we remove this commented code?

File: .github/workflows/parse_snyk_output.cjs
Comment (oweingrod-coforma) [inline .github/workflows/parse_snyk_output.cjs:56]: Nit. Can we make this a single line comment

File: .github/workflows/parse_snyk_output.cjs
Comment (oweingrod-coforma) [inline .github/workflows/parse_snyk_output.cjs:6]: Can we remove this line?

File: .github/workflows/parse_snyk_output.cjs
Comment (oweingrod-coforma) [inline .github/workflows/parse_snyk_output.cjs:27]: Do we need to log vulnerabilities here? I'd prefer to have as little result logging as possible.

File: .github/workflows/parse_snyk_output.cjs
Comment (samzsam) [reply .github/workflows/parse_snyk_output.cjs]: Done!

File: .github/workflows/parse_snyk_output.cjs
Comment (samzsam) [reply .github/workflows/parse_snyk_output.cjs]: Done!

File: .github/workflows/parse_snyk_output.cjs
Comment (samzsam) [reply .github/workflows/parse_snyk_output.cjs]: Removed!

File: .github/workflows/parse_snyk_output.cjs
Comment (samzsam) [reply .github/workflows/parse_snyk_output.cjs]: Done!

**Review by oweingrod-coforma — DISMISSED** [review]

**Comment by oweingrod-coforma** [comment]
Thanks @samzsam! LGTM, feel free to squash merge whenever you want.

**Review by oweingrod-coforma — APPROVED** [review]

**Pull request merged**
